### PR TITLE
Calc perf

### PR DESCRIPTION
**Pull request by pld**
@pld please review our changes, these are mainly changes to the way dframe behaves, e.g. during calculation bamboo does not require the entire dataframe in memory, but only the relevant columns for the calculation, it will then update rows accordingly

(replacing https://github.com/pld/bamboo/pull/46)

### CODE REVIEW

File: bamboo/controllers/calculations.py [hunk 1]
```diff
@@ -82,6 +82,7 @@ def action(dataset):
                 success_message = 'created calculation: %s for dataset: %s' % (
                     name, dataset_id)
 
+
```
Comment (pld) [inline bamboo/controllers/calculations.py]: remove, not-pep8

File: bamboo/core/calculator.py [hunk 2]
```diff
@@ -62,21 +60,24 @@ def calculate_columns(self, calculations):
...
-                aggregation = self.parse_aggregation(
+                aggregator = self.parse_aggregation(
                     c.formula, c.name, c.groups_as_list)
-                aggregation.save()
+                aggregator.save()
             else:
-                columns = self.parse_columns(c.formula, c.name)
-                new_dframe = new_dframe.join(columns[0])
+                columns = self.parse_columns(c.formula, c.name,
+                        length=self.dataset.num_columns)
```
Comment (pld) [inline bamboo/core/calculator.py]: pep8 it

File: bamboo/core/calculator.py [hunk 3]
```diff
@@ -62,21 +60,24 @@ def calculate_columns(self, calculations):
...
+                if new_cols is None:
+                    new_cols = DataFrame(columns[0])
+                else:
+                    new_cols = new_cols.join(columns[0])
 
-        if len(new_dframe.columns) > len(self.dframe.columns):
-            self.dataset.replace_observations(new_dframe)
+                # TODO update instead of replace observations
```
Comment (pld) [inline bamboo/core/calculator.py]: this looks done, or push down to update_obs method

File: bamboo/core/calculator.py [hunk 4]
```diff
@@ -172,24 +173,49 @@ def calculate_updates(self, new_data, new_dframe_raw=None,
...
+
+        # make select from dependent_columns
+        if dframe is None and dependent_columns:
+            query_args = QueryArgs(select={col: 1 for col in dependent_columns})
+            dframe = self.dataset.dframe(query_args=query_args,
+                keep_mongo_keys=True).set_index('MONGO_RESERVED_KEY_id')
+        elif dframe is None:
+            # constant column, use dummy
+            query_args = QueryArgs(select={'_id': 1})
+            dframe = self.dataset.dframe(query_args=query_args,
```
Comment (pld) [inline bamboo/core/calculator.py]: looks like you can define `query_args` in the conditional then pull the `dframe` outside of it

File: bamboo/core/calculator.py [hunk 5]
```diff
@@ -172,24 +173,49 @@ def calculate_updates(self, new_data, new_dframe_raw=None,
...
+                    keep_mongo_keys=True).set_index('MONGO_RESERVED_KEY_id')
+            dframe['dummy'] = 0
```
Comment (pld) [inline bamboo/core/calculator.py:212]: why is this here?

File: bamboo/core/calculator.py [hunk 6]
```diff
@@ -172,24 +173,49 @@ def calculate_updates(self, new_data, new_dframe_raw=None,
...
         """Parse formula into function and variables."""
-        if dframe is None:
-            dframe = self.dataset.dframe()
-
-        functions = self.parser.parse_formula(formula)
+        functions, dependent_columns = self.parser.parse_formula(formula)
+
+        # make select from dependent_columns
+        if dframe is None and dependent_columns:
+            query_args = QueryArgs(select={col: 1 for col in dependent_columns})
```
Comment (pld) [inline bamboo/core/calculator.py]: this is some hotness

File: bamboo/core/frame.py [hunk 7]
```diff
@@ -35,16 +35,20 @@ def add_parent_column(self, parent_dataset_id):
...
         rename_dict = {}
 
         for key in reserved_keys:
-            del self[key]
-            prefixed_key = mongo_prefix_reserved_key(key)
-            if prefixed_key in self.columns:
-                rename_dict[prefixed_key] = key
+            if keep_mongo_keys:
+                replacement_key = MONGO_RESERVED_KEY_PREFIX + key
+                self.__swap_column_names(key, replacement_key)
```
Comment (pld) [inline bamboo/core/frame.py]: why not use:

``` python
df.rename(columns={key: replacement_key, replacement_key: key}, inplace=True)
```

this works fine in ipython

File: bamboo/core/frame.py [hunk 8]
```diff
@@ -91,6 +95,16 @@ def to_csv_as_string(self):
         self.to_csv(buffer, encoding='utf-8', index=False)
         return buffer.getvalue()
 
+    def __column_intersect(self, _list):
+        """Return the intersection of `_list` and this DataFrame's columns."""
+        return list(set(_list).intersection(set(self.columns.tolist())))
+
+    def __swap_column_names(self, column_1, column_2):
```
Comment (pld) [inline bamboo/core/frame.py]: think you can remove this method

File: bamboo/core/merge.py [hunk 9]
```diff
@@ -32,6 +32,10 @@ def merge_dataset_ids(dataset_ids, mapping):
 
     new_dataset = Dataset.create()
 
+    # XXX not sure if we need to do this...
```
Comment (pld) [inline bamboo/core/merge.py]: can we see what happens?  or make this a TODO

File: bamboo/core/operations.py [hunk 10]
```diff
@@ -7,6 +7,7 @@
 
 from bamboo.lib.datetools import safe_parse_date_to_unix_time,\
     parse_str_to_unix_time
+from bamboo.lib.query_args import QueryArgs
```
Comment (pld) [inline bamboo/core/operations.py:10]: yeah!  glad to see the liberal use of QueryArgs

File: bamboo/core/operations.py [hunk 11]
```diff
@@ -29,27 +30,31 @@ def operator_operands(self, tokenlist):
...
+        return []
+
+    def dependent_columns(self, context):
+        return []
+
 
 class EvalConstant(EvalTerm):
     """Class to evaluate a parsed constant or variable."""
 
+    def __cast_to_float(self, value):
```
Comment (pld) [inline bamboo/core/operations.py]: the `lib/utils.parse_float` function almost does this.  this is a generally useful function, can we move to `lib/utils` and combine with `parse_float`?  I think parsing to an `float64` in all cases should be fine, if not we can add a `parse_float64`

File: bamboo/core/calculator.py (hunk 6 above)
Comment (prabhasp) [inline bamboo/core/calculator.py]: Yeah, holy smoking 2-liner, dudes.

File: bamboo/core/operations.py [hunk 12]
```diff
@@ -29,27 +30,31 @@ def operator_operands(self, tokenlist):
...
-                context.dependent_columns.add(self.value)
 
             # test is date and parse as date
             return self.__parse_field(field, context)
 
     def __parse_field(self, field, context):
-            schema = context.schema
-
-            if schema and schema.is_date_simpletype(self.value):
+            if context.dataset and context.dataset.schema.is_date_simpletype(self.value):
```
Comment (pld) [inline bamboo/core/operations.py]: I prefer breaking out schema, it doesn't violate the Law of Demeter, https://en.wikipedia.org/wiki/Law_of_Demeter#In_object-oriented_programming, could write as

``` python
scheme = context.dataset.schema if context.dataset else None
if schema and schema.is_date_simpletype(self.value): 
```

which is better, but still violating, a future refactor might fix this by removing the context notion (for another branch).

File: bamboo/core/operations.py [hunk 13]
```diff
@@ -57,6 +62,14 @@ def __parse_field(self, field, context):
     def field(self, row):
         return row.get(self.value)
 
+    def dependent_columns(self, context):
+        # if value is not number or date, add as a column
+        try:
```
Comment (pld) [inline bamboo/core/operations.py]: As noted above, this try logic should be captured through `lib/utils.parse_float` or a similar function that you could add there and build on `lib/utils._parse_type`.

File: bamboo/core/operations.py [hunk 14]
```diff
@@ -98,6 +114,12 @@ def eval(self, row, context):
 
         return result
 
+    def get_children(self):
+        children = [self.value[0]]
```
Comment (pld) [inline bamboo/core/operations.py:118]: maybe cleaner/faster as a list iteration,

``` python
return [self.value[0]] + [val for oper, val in self.operator_operands(self.value[1:])]
```

or clearer

``` python
first, rest = [self.value[0]], self.value[1:]
return first + [val for oper, val in self.operator_operands(rest)]
```

are there more semantic names for `first` and `rest`?  what makes `rest` special?

File: bamboo/core/operations.py [hunk 15]
```diff
@@ -171,6 +202,12 @@ def eval(self, row, context):
...
```
Comment (pld) [inline bamboo/core/operations.py:206]: with the above, looks like a method is in order.  call as

``` python
extract_children(self.value, self.operator_operands)
```

could go at the module level

File: bamboo/core/operations.py [hunk 16]
```diff
@@ -218,6 +259,11 @@ def eval(self, row, context):
 
         return False
 
+    def get_children(self):
+        if self.tokens[0] == 'default':
```
Comment (pld) [inline bamboo/core/operations.py:264]: ah yes, the special MapOp, the string `'default'` is special because it's also special in the syntax of MapOp, maybe a comment to make this clear?

or a special value also replaces any string reference to `'default`' in the parser.

File: bamboo/core/operations.py [hunk 17]
```diff
@@ -239,6 +291,14 @@ class EvalPercentile(EvalFunction):
 
     def eval(self, row, context):
         # parse date from string
-        column = context.dframe[self.value.value]
+        col = self.value.value
+        query_args = QueryArgs(select={col: 1})
+        column = context.dataset.dframe(query_args=query_args)[col]
```
Comment (pld) [inline bamboo/core/operations.py:299]: boom!  why get a dframe when you need a column, that's a damn good motto

File: bamboo/core/parser.py [hunk 18]
```diff
@@ -19,12 +19,9 @@ class ParserContext(object):
     """Context to be passed into parser."""
 
     dependent_columns = set()
-    schema = None
 
     def __init__(self, dataset=None):
```
Comment (pld) [inline bamboo/core/parser.py:23]: I'm now of the opnion that the whole ParserContext was a bad idea (whoops), making an issue to get rid of it, let's just pass dataset through and worry about needed more objects when the time comes

Comment (pld) [inline bamboo/core/parser.py:23]: https://github.com/modilabs/bamboo/issues/515

File: bamboo/core/parser.py [hunk 19]
```diff
@@ -294,21 +296,30 @@ def validate_formula(self, formula, row):
         self.aggregation = None
 
         # check valid formula
-        functions = self.parse_formula(formula)
+        functions, dependent_columns = self.parse_formula(formula)
 
-        if not self.context.schema:
+        if not self.context.dataset or not self.context.dataset.schema:
```
Comment (pld) [inline bamboo/core/parser.py]: shouldn't this be `and`?  if we don't have `self.context.dataset` then second disjunct will raise an `AttributeError`.

File: bamboo/core/summary.py [hunk 20]
```diff
@@ -80,11 +80,17 @@ def summarize(dataset, dframe, groups, no_cache):
...
     group_stats = stats.get(group_str)
 
-    if no_cache or not group_stats:
+    if no_cache or not group_stats or update:
         group_stats = summarize_with_groups(dframe, groups, dataset) if\
             groups else summarize_df(dframe, dataset=dataset)
 
         if not no_cache:
+            if update:
+                #adding the new group_stats back to the old stat
```
Comment (pld) [inline bamboo/core/summary.py]: formatting

File: bamboo/core/summary.py [hunk 21]
```diff
@@ -80,11 +80,17 @@ def summarize(dataset, dframe, groups, no_cache):
...
+                original_group_stats = stats.get(group_str, {})
+                original_group_stats.update(group_stats)
```
Comment (pld) [inline bamboo/core/summary.py]: I'd suggest this simple (well, could be simpler) performant pattern:

``` python
from itertools import chain

group_stats = dict(chain(original_group_stats.iteritems(), group_stats.iteritems())
```

Better yet, we could put this in a `lib/utils` helper method, I believe we're doing the same thing in other places in the code

File: bamboo/lib/query_args.py [hunk 22]
```diff
@@ -34,3 +34,7 @@ def __init__(self, query={}, select=None, distinct=None, limit=0,
         self.distinct = distinct
         self.limit = limit
         self.order_by_value = order_by
+
+    def __nonzero__(self):
```
Comment (pld) [inline bamboo/lib/query_args.py:38]: good stuff.

File: bamboo/models/abstract_model.py [hunk 23]
```diff
@@ -137,6 +137,24 @@ def command(records):
         cls.__batch_command_wrapper(command, dframe, batch_size)
 
     @classmethod
+    def batch_update(cls, dframe):
+        """Update records in batches to avoid document size maximum setting.
+           dframe must have column with record (object) ids.
+
+        """
+        def command(records):
+            # mongodb has no batch updates (fail)
```
Comment (pld) [inline bamboo/models/abstract_model.py:146]: super fail

File: bamboo/lib/debug.py [hunk 24]
```diff
@@ -0,0 +1,22 @@
+from datetime import datetime
```
Comment (pld) [inline bamboo/lib/debug.py]: I don't think this should go in the `/bamboo/bamboo` directory if it is not used anywhere in the code.  `/scripts` maybe?  I think we should keep `/bamboo/bamboo` to things that are used by the bamboo appliction during runtime.  In the future this means moving the `mongo_index` up and out, and maybe the tests.

File: bamboo/core/calculator.py (hunk 6 above)
Comment (myf) [inline bamboo/core/calculator.py]: devils advocate here:

though I love one-liners and elegant codes, sometimes I do find very concise codes that does many things in one line hard to debug. just a thought

File: bamboo/models/abstract_model.py [hunk 25]
```diff
@@ -137,6 +137,24 @@ def command(records):
...
+            for record in records:
+                spec = {'_id': record['MONGO_RESERVED_KEY_id']}
+                del record['MONGO_RESERVED_KEY_id']
+                doc = {"$set": record}
```
Comment (pld) [inline bamboo/models/abstract_model.py:150]: IMAGE

File: bamboo/models/calculation.py [hunk 26]
```diff
@@ -69,8 +70,8 @@ def calculate_task(calculations, dataset):
     :param dataset: Dataset to run calculation on.
     """
     # block until other calculations for this dataset are finished
+    # XXX this looks sketchy
```
Comment (pld) [inline bamboo/models/calculation.py]: doesn't it?  can we try putting in the iteration as:

``` python
calculation.restart_if_has_pending(dataset, [c for c in calculations if c != calculation])
```

File: bamboo/models/dataset.py [hunk 27]
```diff
@@ -175,24 +179,36 @@ def aggregated_dataset(self, groups):
...
 
     def dframe(self, query_args=QueryArgs(), keep_parent_ids=False,
-               padded=False, index=False):
+               padded=False, index=False, reload=False, keep_mongo_keys=False):
         """Fetch the dframe for this dataset.
 
         :param query_args: An optional QueryArgs to hold the query arguments.
         :param keep_parent_ids: Do not remove parent IDs from the dframe,
             default False.
+        :param padded: Used for joining, default False.
```
Comment (pld) [inline bamboo/models/dataset.py:185]: always appreciated to get the docs up to date

File: bamboo/core/calculator.py (hunk 6 above)
Comment (pld) [inline bamboo/core/calculator.py]: this is pretty straightfoward, but agreed, maybe cleaner as:

``` python
column_select = {col: 1 for col in dependent_columns}
query_args = QueryArgs(select=column_select)
```

File: bamboo/models/dataset.py [hunk 28]
```diff
@@ -175,24 +179,36 @@ def aggregated_dataset(self, groups):
...
         :param index: Return the index with dframe, default False.
+        :param reload: Force refresh of data, default False.
+        :param keep_mongo_keys: Used for updating documents, default False.
 
         :returns: Return BambooFrame with contents based on query parameters
             passed to MongoDB. BambooFrame will not have parent ids if
             `keep_parent_ids` is False.
         """
+        cacheable = True
```
Comment (pld) [inline bamboo/models/dataset.py]: for me this would be simpler and easier to understand as:

``` python
cacheable = not (query_args or keep_parent_ids or padded)
```

File: bamboo/models/dataset.py [hunk 29]
```diff
@@ -51,6 +51,10 @@ class Dataset(AbstractModel, ImportableDataset):
     SCHEMA = 'schema'
     UPDATED_AT = 'updated_at'
 
+    def __init__(self, record=None):
+        super(Dataset, self).__init__(record)
+        self._dframe = None
```
Comment (pld) [inline bamboo/models/dataset.py]: Do we want `_dframe` to be only accessed within the class?  We should use two underscores, so `self.__dframe`, http://docs.python.org/release/1.5.1p1/tut/private.html, the classic "private" instance variable.

File: bamboo/models/observation.py [hunk 30]
```diff
@@ -89,6 +107,7 @@ def save(cls, dframe, dataset):
         else:
             cls.batch_save(dframe)
 
+        # XXX do we need this? (merge)
```
Comment (pld) [inline bamboo/models/observation.py]: I don't think we do.  Do tests pass without it?  Let's check and remove both lines if they do.

File: bamboo/tests/controllers/test_abstract_datasets.py [hunk 31]
```diff
@@ -56,6 +56,7 @@ def _check_dframes_are_equal(self, dframe1, dframe2):
         self._check_dframe_is_subset(dframe2, dframe1)
 
     def _check_dframe_is_subset(self, dframe1, dframe2):
+        from pandas import DataFrame
```
Comment (pld) [inline bamboo/tests/controllers/test_abstract_datasets.py]: move to top of module

File: bamboo/tests/core/test_calculator.py [hunk 32]
```diff
@@ -43,7 +43,8 @@ def _test_calculator(self):
 
         for idx, formula in enumerate(self.calculations):
             name = 'test-%s' % idx
-            self.parser.validate_formula(formula, row)
+            # XXX do we need this validation?
+            #self.parser.validate_formula(formula)
```
Comment (pld) [inline bamboo/tests/core/test_calculator.py]: This is good to have, this will verify that `validate_formula` does not raise a ParseError, which it should not.

File: bamboo/tests/core/test_parser.py [hunk 33]
```diff
@@ -1,23 +1,26 @@
...
-        self.row = {'VAR': 1}
         TestBase.setUp(self)
+        self.dataset_id = self._post_file()
+        self.dataset = Dataset.find_one(self.dataset_id)
+        self.parser = Parser(self.dataset)
+        self.row = {'amount': 1}
 
     def _check_func(self, parse_result):
-        functions = parse_result
+        functions = parse_result[0]
```
Comment (pld) [inline bamboo/tests/core/test_parser.py]: prefer

``` python
functions, _ = parse_result
```

to ignore second argument

File: bamboo/tests/core/test_parser.py [hunk 34]
```diff
@@ -26,9 +29,20 @@ def test_bnf(self):
 
     def test_parse_formula_with_var(self):
         func = self._check_func(
-            self.parser.parse_formula('VAR + 1'))
+            self.parser.parse_formula('amount + 1'))
```
Comment (pld) [inline bamboo/tests/core/test_parser.py]: Hmm, can we move `self.parser.parse_formula` into `_check_func` and just pass the string to parse into it?

We might want to rename it, `_check_func` => `_parse_and_check` or something

File: bamboo/tests/core/test_parser.py [hunk 35]
```diff
@@ -26,9 +29,20 @@ def test_bnf(self):
...
         self.assertEqual(func(self.row, self.parser.context), 2)
 
+    def test_parse_formula_dependent_columns(self):
+        test_formulae = {
+            '1': set([]),
+            'amount + 1': set(['amount']),
+            'amount + gps_alt': set(['amount', 'gps_alt']),
+            # TODO add aggregations?
```
Comment (pld) [inline bamboo/tests/core/test_parser.py:38]: prolly, how about we have a common set of calculations to test among the test classes?  We can base it on `tests/core/test_calculations.calculations`.

**Comment by pld** [comment]
This is some solid stuff, great work guys!

File: bamboo/core/calculator.py (hunk 5 above)
Comment (pld) [inline bamboo/core/calculator.py:212]: for the count eh, hmm... can we pull out some arbitrary column of the dframe (by adding it to the select), perhaps a reserved column we know will exist?

**Comment by mejymejy** [comment]
yeah, some of the code is still a bit dirty... just wanted to get it up here for comments, etc.  still need to pep8 and clean it up a bit.  will update.

File: bamboo/lib/debug.py (hunk 24 above)
Comment (myf) [inline bamboo/lib/debug.py]: it seems like a good idea to integrate mongo_index as the initial set up on a server or an empty db. it is relevant to bamboo and it sped up bamboo operation hundreds folds.

Comment (mejymejy) [inline bamboo/lib/debug.py]: i think that it needs to be run on any install, but should probably be pulled out and put into scripts and added to the documentation outlining initial server setup.  not sure if there is a way to guarantee that it happens (it can be ran every time the server starts up if necessary since it is idempotent).  we can add it to scripts/run_server.py but not everyone will use it.  therefore, maybe it makes sense to have a standalone script to run, but one that can also be called from scripts.

Comment (pld) [inline bamboo/lib/debug.py]: let's just put in in the docs, running it everytime the server starts would make starting the server slower.

Comment (myf) [inline bamboo/lib/debug.py]: it is ensure_indexing though, after the first time it should be just fine

Comment (mejymejy) [inline bamboo/lib/debug.py]: i think we should actually move stuff like this print_time function into its own pip-installable library in modilabs repo (call it modilabs-debugtools or something) so that we can use it in multiple projects.  taking it out from bamboo.

**Comment by mejymejy** [comment]
@pld thanks for the detailed code inspection!  i addressed a bunch of the issues and will fix the others later tonight.  i will add issues for those that i think should be addressed outside of this pull request.  hoping to have a merge-ready pr for tomorrow...

**Comment by pld** [comment]
Sounds good, it will be great to get this into master.

File: bamboo/tests/core/test_parser.py (hunk 35 above)
Comment (mejymejy) [inline bamboo/tests/core/test_parser.py:38]: would be nice, opened issue #516 for it for now since we can work on that after this is merged

File: bamboo/models/calculation.py (hunk 26 above)
Comment (mejymejy) [inline bamboo/models/calculation.py]: after looking at this again, i am fine with it.  removing XXX comment

File: bamboo/core/operations.py (hunk 17 above)
Comment (mejymejy) [inline bamboo/core/operations.py:299]: not as good as [this one](http://www.youtube.com/watch?v=W45DRy7M1no&t=2m25s)!

File: bamboo/core/operations.py (hunk 14 above)
Comment (mejymejy) [inline bamboo/core/operations.py:118]: what is special about the first versus the rest is that this type of parent "extraction" is for binary ops that can be chained e.g. "p or q and s" where the first value is there and the rest are packaged from the parser with the operation e.g. [p, (or, q), (and, s)].  i tried to group these parent-finding functions, but they are difficult because they depend on how the parser returns the tokens.  there are essentially 5 groups (right now):
1. no children (`[]`)
2. just the node's value (`[self.value]`)
3. the nodes inside a node's value (`[val for val in self.value]`)
4. this type of extraction for binary stuff
5. the map op special case

i think that this should be refactored, but am not sure on the best way.  i have opened issue #517 so that we can keep this discussion going after we merge this pr in.

File: bamboo/tests/controllers/test_datasets_ts_functions.py [hunk 36]
```diff
@@ -74,8 +74,8 @@ def test_resample_bad_interval(self):
         self.assertEqual(
             'Could not evaluate %s' % interval, result[Datasets.ERROR])
 
-    def test_resample_with_query(self):
-        expected_length = 15
+    def test_resample(self):
+        expected_length = 17
```
Comment (mejymejy) [inline bamboo/tests/controllers/test_datasets_ts_functions.py:78]: @pld can you confirm that this is correct? this test was inadvertently overrided by the test below it and therefore was not running.  i think i fixed it correctly...

**Comment by mejymejy** [comment]
@pld @myf @prabhasp @rgaudin, i think this is ready to go!  @pld can you check [this](https://github.com/modilabs/bamboo/pull/513/files#L25R77) just to make sure it's correct?

File: bamboo/tests/controllers/test_datasets_ts_functions.py (hunk 36 above)
Comment (pld) [inline bamboo/tests/controllers/test_datasets_ts_functions.py:78]: @mejymejy yep, looks good

**Comment by pld** [comment]
do it!
